### ISSUE TITLE
[A11y]: component elements do not have programmatically determinable name for announcements 

### ISSUE THREAD

**Issue by NatalliaAlieva** [epam/UUI#1553]
### Description
There is no announcement of role and state and changed state of Accordions

### Precondition
Screen Reader is **ON**

### Steps to Reproduce
1. Open [Accordion component](https://uui.epam.com/documents?id=accordion&mode=doc&skin=UUI4_promo&category=components).
2. Using `Tab` focus any of the presented Accordions examples.
3. Using `Enter`/`Space` key expand or collapse the section and note the announcement. 

### Actual Result
"[Name], group" is announced (e.g. Accordion block mode, group). There is no announcement of role and state. There is no announcement of changed state after pressing Enter/Space key. 

https://github.com/epam/UUI/assets/57751283/37b1b2f4-142d-4d11-9561-fdafc066ec0c

### Expected Result
Screen reader should announce accordion header name, role "button" and current state "collapsed" or "expanded" (e.g. Accordion block mode, collapsed, button). 
After pressing Space/Enter key screen reader should announce the changed state "collapsed" or "expanded".
Reference for implementation: https://www.w3.org/WAI/ARIA/apg/patterns/accordion/examples/accordion/
**Failed WCAG Checkpoint:** 4.1.2 Name, Role, Value (Level A) 
https://www.w3.org/TR/WCAG21/#name-role-value

### Description
Plain text "Employee" on Custom accordion is announced as a button by screen reader

### Precondition
Screen Reader is **ON**

### Steps to Reproduce
1. Open [Accordion component](https://uui.epam.com/documents?id=accordion&mode=doc&skin=UUI4_promo&category=components).
2. Use the `down arrow` key to navigate to "Employee" text on "Custom accordion" and note the announcement. 

### Actual Result
Plain text "Employee" on "Custom accordion" is announced as a button by screen reader.
![image](https://github.com/epam/UUI/assets/57751283/26b3065d-a6c3-4a57-9f9c-0bada29ea53f)

### Expected Result
Plain text "Employee" on "Custom accordion" should not be announced as an interactive element button by screen reader.

**Failed WCAG Checkpoint:** 4.1.2 Name, Role, Value (Level A) 
https://www.w3.org/TR/WCAG21/#name-role-value

**Comment by vasilii-kovalev** [comment]
After discussing the case 37 with UUI team, it was decided that it won't be fixed. Below you'll find an explanation why.

## Current solution

`MainMenu` component is wrapped in a `header` element (a user have to do that manually). `MainMenu` contains a `nav` element as well as `AdaptivePanel` inside it, which renders two `div` elements inside one another. Then the user's elements are rendered inside the last `div` element. So the structure looks like this:

```html
<header>
  <nav>
    <div>
      <div>
        <a href="...">Link 1</a>
        <a href="...">Link 2</a>
        ...
      <div>
    <div>
  </nav>
</header>
```

The `div`-s presence doesn't make much of a difference, so they will be omitted in the further examples.

## Better solutions

It is necessary to separate elements, related to navigation, like list of links and buttons, that show/hide a list of links, from other elements, like logotype (even if wrapped in a link), buttons, that do not show/hide a list of links, inputs, and so on. It is necessary to do so mostly by the meaning of the content, though there are no clear recommendations from WCAG or WAI-ARIA about that. But at least it is clear, how to organize the elements, related to navigation[^1][^2]. The structure should look like this:

```html
<header>
  <nav>
    <ul>
      <li>
        <a href="...">Link 1</a>
      </li>
      <li>
        <a href="...">Link 2</a>
      </li>
      ...
    </ul>
  </nav>
</header>
```

We've discussed 3 options of such mixed content organization with an accessibility expert within the company. Here's the results.

### Logotype in the `ul` element

The proposition is to create a list item in the list and put the logotype in it.

Example:

```html
<header>
  <nav>
    <ul>
      <li>
        <a href="...">
          <img src="..." alt="Logotype">
        </a>
      </li>
      <li>
        <a href="...">Link 1</a>
      </li>
      <li>
        <a href="...">Link 2</a>
      </li>
      ...
    </ul>
  </nav>
</header>
```

It is considered the worst option, because it combines all types of content in a place, that should be solely dedicated to navigation by its meaning. Screen readers will announce the amount of all links and buttons in the list, even if the buttons do not show/hide a list of links.
Let's call it "the worst solution" for further references.

### Logotype in the `nav` element

The proposition is to put the logotype outside the `ul` element, but inside the `nav` element.

Example:

```html
<header>
  <nav>
    <a href="...">
      <img src="..." alt="Logotype">
    </a>
    <ul>
      <li>
        <a href="...">Link 1</a>
      </li>
      <li>
        <a href="...">Link 2</a>
      </li>
      ...
    </ul>
  </nav>
</header>
```

This solution is better than the previous one, and can be found more commonly on the Internet[^3][^4]. But the logotype and other such elements are still not a part of the navigation by their meaning, so this solution is not ideal.
Let's call it "the acceptable solution" for further references.

### Logotype in the `header` element

The proposition is to put the logotype outside the `nav` element, but inside the `header` element.

Example:

```html
<header>
  <a href="...">
    <img src="..." alt="Logotype">
  </a>
  <nav>
    <ul>
      <li>
        <a href="...">Link 1</a>
      </li>
      <li>
        <a href="...">Link 2</a>
      </li>
      ...
    </ul>
  </nav>
</header>
```

This solution is considered the best of all, because it clearly separated navigation links from other elements.
Let's call it "the best solution" for further references.

##  Applying the better solutions to real-world cases

`AdaptivePanel`, which is used inside the `MainMenu` component, doesn't know what elements are present in the `items` array, so it can't automatically separate navigation-related elements from others. This knowledge exists on the user's side. Also, keeping the functionality of atomic visibility control for each item is crucial, and won't be changed just for the sake of accessibility compliance.

With that in mind, here's what UUI would have to do to implement the better solutions.

### Implementing the best solution

To implement that solution, it would be necessary to remove `nav` from `MainMenu` and tell users to wrap the navigation links in it and in `ul`.

#### Pros

1. Accessibility is driven by users and as perfect as they want it to be

#### Cons

1. More effort from the user
2. The atomic control is gone - the whole navigation links block will be shown/hidden instead of each link in it

Because of the latter problem, the solution can't be implemented.

### Implementing the acceptable solution

To implement that solution, it would be necessary to keep `nav` in the `MainMenu`, but still tell users to wrap the navigation links in `ul`.

The cons are the same, as for the best solution, but accessibility now is a bit worse, so the solution can't be implemented.

### Implementing the worst solution

To implement that solution, it would be necessary to set `role="list"` to `AdaptivePanel`, and tell users to wrap the navigation links in `li`.

#### Pros

1. The atomic control is preserved

#### Cons

1. More effort from the user
2. The list (controlled by UUI) and its items (controlled by users) are disconnected. Users have to assume UUI still implements the list in the future
3. Accessibility "improvements" are questionable, because non-navigation elements are included in the list and there is no way for the user to opt-out without additional code to remove the `role` from the `AdaptivePanel`, if they decided to improve accessibility on their side

After discussing this option with the accessibility expert, is was decided that it is better to keep the code unchanged, since it'll make experience for screen reader users worse.

## Other ideas

### Components composition

Instead of setting elements in the `items` array, `MainMenu` (and `AdaptivePanel`) could accept children, so that users could define the structure they want. It would look like this:

```jsx
<header>
  <MainMenu>
    <ul>
      <li>
        <a href="...">Link 1</a>
      </li>
      <li>
        <a href="...">Link 2</a>
      </li>
      ...
    </ul>
  </MainMenu>
</header>
```

But this approach has a significant problem: `AdaptivePanel` can't connect to the children to measure their dimensions and use atomic control without using `React.Children`, which UUI will stop using at some point, because it leads to a fragile code, which is mentioned in React documentation[^5].

### Adding `listitem` role to `MainMenuButton`

When implementing "the worst solution", instead of asking users to wrap each item in `li` elements, it was proposed to do it on UUI side and add `role="listitem"` to `MainMenuButton` component. This approach leads to a conflict of roles. For example, a link variant, which previously would be announced as "Github repo visited link", with the manually assigned role would be announced as "clickable", ignoring `aria-label` attribute, which is unacceptable.

### Using multiple `AdaptivePanel` instances

To be able to keep the atomic control, when user wraps the item passed to `MainMenu`, in `ul`, it was proposed to wrap the list items in `AdaptivePanel`. That would theoretically solve the problem, but introduce another - it would be hard or even impossible to send information about the hidden/shown items to the main `AdaptivePanel` (inside the `MainMenu` component).

## References

[^1]: [ARIA Landmark Example](https://www.w3.org/WAI/ARIA/apg/patterns/landmarks/examples/navigation.html)
[^2]: [Fly-out Menus](https://www.w3.org/WAI/tutorials/menus/flyout)
[^3]: [W3C](https://www.w3.org)
[^4]: [Bootstrap](https://getbootstrap.com)
[^5]: [Children – React](https://react.dev/reference/react/Children)

**Comment by mari-shakhanava** [comment]
### **Case №45**
**Description**
[File Card]: Name and state of 'delete file' button is not announced by screen reader

**Precondition**
Screen Reader is ON

**Steps to Reproduce**
1. Open [File Upload](https://uui.epam.com/documents?id=fileUpload&mode=doc&category=components&theme=loveship) component.
2. Upload any file.
3. Use Tab key to navigate to X button in the File Card and note the announcement.

**Actual Result**
Name of the button is not announced by screen reader, only "button" is announced.

![Image](https://github.com/user-attachments/assets/a6e6bb44-8b5f-47d9-8a6c-23a987fb86ff)

**Expected Result**
Name (e.g. "delete file")  of the button should be announced by screen reader.
Failed WCAG Checkpoint: 4.1.2 Name, Role, Value (Level A)
https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html

**Comment by NatalliaAlieva** [comment]
**_Comment for case 12:_**

On macOS VoiceOver, selected options are still announced as “not selected”.

https://github.com/user-attachments/assets/f2ab41cc-e1dc-4264-8b32-2d661e66ee1d

On Windows NVDA, the same action is announced correctly (“checked”).

https://github.com/user-attachments/assets/64757273-4035-49db-b1a5-fe77d31d1cd1

Differences in announcement are screen-reader-specific and not caused by the component logic.

**Comment by MSt1ch** [comment]
Fixed **Case №1**
